pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java

```diff
@@ -210,14 +210,22 @@ public void publishMessage(long producerId, long lowestSequenceId, long highestS
 
     public boolean checkAndStartPublish(long producerId, long sequenceId, ByteBuf headersAndPayload, long batchSize,
                                         Position position) {
-        if (isShadowTopic && position == null || !isShadowTopic && position != null) {
+        if (!isShadowTopic && position != null) {
             cnx.execute(() -> {
                 cnx.getCommandSender().sendSendError(producerId, sequenceId, ServerError.NotAllowedError,
                         "Only shadow topic supports sending messages with messageId");
                 cnx.completedSendOperation(isNonPersistentTopic, headersAndPayload.readableBytes());
             });
             return false;
         }
+        if (isShadowTopic && position == null) {
+            cnx.execute(() -> {
+                cnx.getCommandSender().sendSendError(producerId, sequenceId, ServerError.NotAllowedError,
+                        "Cannot send messages to a shadow topic");
```

Comment:
We'd better tell this error is caused by `position == null`

Reply:
This `position` is only used in shadow topic internal replication:

https://github.com/apache/pulsar/blob/5b062f37bd6688c09384abbe908f7c14e28052fc/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java#L1598-L1602

`position` should be `null` for regular producers. So I think it's not necessary to tell users about this detail.

Reply:
> `position` should be `null` for regular producers.

Correct. It's only used in `ShadowReplicator`.